### DISCUSSION TITLE
Scope & utility of Magma orc8r RPC segregation effort

### DISCUSSION THREAD

**Discussion by emakeev** [magma/magma#11205] [General]
There are few new PRs targeting RPC segregation. We should probably discuss the effort goals and approaches since it seems a bit disruptive, it moves lots of service files around, reaches into independant Magma modules (FeG, LTE), rewrites established protos & may create maintenance confusion in the future. 

The [PRs in question](https://github.com/magma/magma/issues?q=segregation+in%3Atitle%2Cbody) are: #11186, #11182, #11180, #11178, #11177, #11176, #11175, #11128, #11119, #11105, #11099, #11086, #11079, #11077, #11058, #11047, #11015, #10998, #10997, #10973, #10965, #10962, #10954, #10952, #10950, #10936, #10923, #10917, #10907, #10905, #10902, #10900, #10899, #10898, #10879, #10878, #10875, #10861, #10859, #10855, #10850, #10787, #10784, #10777, #10776, #11652  (some may be duplicates).

As far as I understand from [this document](https://drive.google.com/file/d/1NO6Qd6rU80xPh0Sb0mKWlSfDt0JbeVIT/view), the goal is to limit access to internal only RPCs. 

We seem to be concerned about orc8r efficiency & scalability, yet - we propose to nearly double number of orc8r gRPC services. The codebase disruption of the proposed implementation also appears to be pretty significant (see PRs mentioned above).

Should we consider a lighter, less disruptive & costly solution? I provided some possible alternatives with their rough dev estimates below to demonstrate the scope of the required effort.

Existing magma access control infrastructure already provides means for every RPC to query if the call is internal or external by a registered Gateway. We can leverage this capability to offer the following implementation options:

1) Every Magma orc8r service is instantiated with NewOrchestratorService() call, we can convert existing NewOrchestratorService() implementation to create internal only services and introduce a new service instantiation function for external services - New[External/Southbound]OrchestratorService() (estimated 1 day of eng work). 
Then go over all existing services, determine if they are external or internal and change NewOrchestratorService() to New[External/Southbound]OrchestratorService() for services intended to be used externally (estimated 1-2 days of eng work).

2) Another option is to introduce a registry of external services and RPCs and to modify existing security/access control framework to honor this registry. The actual implementation can be staged, so we can get a working hardcode registry in a matter of days and provide a flexible/configurable implementation in a couple of weeks:
  a) codebased, static registry with a map of existing external services & RPCs (est. 3 days)
  b) YML config based registry with a build script assembling global combined YML registry file during build/deployment from orc8r and deployable modules (1-2 weeks)
  c) If needed - use the same YML based registry to generate deployment nginx config/helm templates. This step is optional, the access control for internal/external service RPCs would be fully functional without it. (1 week)
I think, having a single (or module specific) external/southbound configurable registry  could be less disruptive and error prone then trying to enforce a specific directory location of existing or new services, it'd also put less burden on developers and maintainers (reviewers) modifying existing services and especially writing new services. It'll also allow to colocate services RPC in any way a developer sees logical/fit.

While option 1 is much easier to implement, option 2 gives us greater flexibility (it'll allow a single service to safely mix external and internal RPCs) and greater visibility (single registry listing external RPCs). Both solutions would require minimal changes to existing service & protocol implementations (solution 1 would require one line change per service and solution 2 will not require any changes to the existing service implementation code; none of them will require any procol changes).

3) The third option is a subset of option 2 above, we can still rely on a single configurable registry (2.b), but only use it in generation of corresponding deployment nginx config/helm templates (2.c). This approach would work for current Magma deployment models, but may prove to be less portable in the future, while option 2 ensures that access control enforcement is also guaranteed by built into every service access control layer, option 3's enforcement would solely rely on external to Magma components.

**Comment by hcgatewood** [comment]
Hey @emakeev! Thanks for the detailed writeup. To assuage your concerns, I'll note that we're not actually creating any new gRPC services [*], and instead are just (a) clarifying which servicers are internal vs. external, and (b) moving the internal-only services to their own, dedicated internal-only, port [**]

This should have, net, no effect on codebase complexity, and is conceptually a straightforward solution

Happy to chat over it further if desired 🙂 

[*] The only asterisk here is we have a couple gRPC servicers that are "mixed", i.e. some endpoints are external and some are internal. In those cases we will indeed have to add new servicers, but -- there are only 4 cases like this, meaning we're adding 4 new servicers to a codebase currently with 35+
[**] The benefit of this solution is we get network-level access enforcement, rather than application-level (or, servicer-level rather than servicer endpoint-level, depending on how you wanna think about it) -- and we'll force each servicer creation to choose whether it wants to be external or internal, upon startup. Imo this extreme clarity is a win in security-critical situations.

**Reply by emakeev** [reply]
Hi @hcgatewood. How do you see the maintenance & enforcement of this external/internal service separation? When developers add new services, will they be required to use specific directory paths/names? And if yes, how is it going to be enforced? Through PR reviews or via some automated check on PR submission? 

I just thought that having a single (or at least a single per module) list of permitted external services & RPC (for mixed services) would be easier to maintain and understand, esp. for developers new to Magma. Also, even properly moving large pieces of service implementations to a new directory could make parsing implementation/feature/bug history a bit harder (still possible, but harder).

We can still have the network-level enforcement of this list if it's preferable, the mechanisms that I described in the original proposal are just an easy implementation proposals, the list itself can be used in various ways (including network-level enforcement).

Having such a single point of reference could make life easier for developers & reviewers. Developers would be able to create their services in freer form, without need to split their proto definitions and services and maintain convention enforced directory structure. Reviewers would clearly see a list of new added external services/RPCs and new potential security risks (either a service/RPC is in the list and therefore it's external or it's not in the list and it's internal only).

**Comment by EngineeringBoy** [comment]
邮件已收到，我会尽快回复！
